react/Product.js

```diff
@@ -5,19 +5,45 @@ import PropTypes from 'prop-types'
 // eslint-disable-next-line no-restricted-imports
 import { pathOr, path, sort, last, flatten } from 'ramda'
 import { jsonLdScriptProps } from 'react-schemaorg'
+import { useQuery } from 'react-apollo'
+import GET_SETTINGS from './queries/getSettings.graphql'
 
 const getSpotPrice = path(['commertialOffer', 'spotPrice'])
 const getPrice = path(['commertialOffer', 'Price'])
+const getTax = path(['commertialOffer', 'Tax'])
 const getAvailableQuantity = pathOr(0, ['commertialOffer', 'AvailableQuantity'])
+const DEFAULT_DECIMALS = 2
+const DEFAULT_PRICES_WITH_TAX = false
+
+const getAppSettings = () => {
+  const { data } = useQuery(GET_SETTINGS, { ssr: false })
```

Comment:
Hooks can't be used outside the react tree and the order that hooks are called, matters.

Reply:
Capitalized it to be able to use it as described in https://reactjs.org/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized, I see no other way around it. I need it there to be able to also mock it and test it (not the query itself, but the function that uses its response.

Reply:
Sorry, but this seems like cheating to me 😅 . I really don't think this is the best approach. 

If the only reason you cannot call this hook inside of the `StructuredData` component is because of testing tools, then there's something wrong with the testing tools. 

Buuut, in general, using `@vtex/test-tools`, you should be able to mock and test GraphQL queries without any issues. We do it in this app for example: https://github.com/vtex-apps/product-gifts/blob/master/react/__tests__/ProductGifts.test.tsx. Check out how we create this `renderWithProductContext` utility function: https://github.com/vtex-apps/product-gifts/blob/22fb0e4ad55604c609b3621b1d4e29c9135a247a/react/__mocks__/testUtils.tsx#L17. Something similar to this should work here :)

Reply:
> Sorry, but this seems like cheating to me  . I really don't think this is the best approach.
> 
> If the only reason you cannot call this hook inside of the `StructuredData` component is because of testing tools, then there's something wrong with the testing tools.
> 
> Buuut, in general, using `@vtex/test-tools`, you should be able to mock and test GraphQL queries without any issues. We do it in this app for example: https://github.com/vtex-apps/product-gifts/blob/master/react/__tests__/ProductGifts.test.tsx. Check out how we create this `renderWithProductContext` utility function: https://github.com/vtex-apps/product-gifts/blob/22fb0e4ad55604c609b3621b1d4e29c9135a247a/react/__mocks__/testUtils.tsx#L17. Something similar to this should work here :)

Hi @victorhmp - I see what you're saying, unfortunately I'm not very familiar with tests in React, only learned it a year ago and before that I only had experience with PHP/Python. I can't figure out a quick solution on how to mock graphQL into the test, so if you're not happy with the "workaround" that passes the test and has no impact on code execution whatsoever (AFAIK), could you please do the changes yourself?  I changed the manifest to accept i18n messages and would help a lot if you could offer some assistance for the test so we could merge this ASAP.   The client's been waiting for a week now...

Thanks

Reply:
This is really an issue. You have to transform it to a hook (`useAppSettings`) and provide a default implementation as a mock for the testing suite. You can make it return a fixed value or have multiple implementations for it depending on if you want to test the tax and decimals behavior.